### PR TITLE
`General`: Improve detailed feedback view for exercises

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The result detail view (when you click on 7 of 13 passed and see the test feedback) gets improved.

### Description

- The text about potentially hidden test feedback after the due date is now only shown before the latest individual due date
- If the static code analysis would deduct more points than is allowed by the settings in the gradings tab, the student correctly sees the deducted amount and an explanatory tooltip.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Student
- 1 of each Exercise. (Programming exercise needs SCA feedback and Submissionpolicy and some manual feedback)

1. Log in to Artemis
2. Create the exercises
3. Participate as a student 
4. Grade the exerises
5. Check the feedback in the student view
6. Set the individual due date of the other student to some later date
7. Check that the student without the longer due date can see a notification about potentially hidden feedback at the top of the feedback view

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| result-detail.component.ts | 87.31% | 94.4% |
| ParticipationService.java | 78% | 94% |
| ProgrammingExerciseGradingService.java | 87% | 96% |

### Screenshots

Bildschirmfoto 2022-06-08 um 21 22 39

### CODE REVIEW

File: src/main/webapp/i18n/de/result.json [hunk 1]
```diff
@@ -74,7 +74,7 @@
             },
             "notLatestSubmission": "Hinweis: Dieses Ergebnis und Feedback bezieht sich <strong>nicht</strong> auf die letzte Einreichung.",
             "seeMore": "Mehr anzeigen",
-            "afterDueDateFeedbackHidden": "Auch nach der Abgabefrist werden nicht immer Detailfeedbacks zu allen Testfällen angezeigt. Dies ist etwa der Fall wenn für andere Studierende eine spätere Abgabefrist gesetzt ist (z.B. zum Ausgleich von Krankheitstagen). Sobald alle Abgaben erfolgt sind, werden auch diese Feedbacks automatisch sichtbar."
+            "afterDueDateFeedbackHidden": "Auch nach der Abgabefrist werden nicht Detailfeedbacks zu allen Testfällen angezeigt. Dies ist etwa der Fall wenn für andere Studierende eine spätere Abgabefrist gesetzt ist (z.B. zum Ausgleich von Krankheitstagen). Alle Feedbacks werden automatisch am {{date}} sichtbar sein."
```
Comment (Strohgelaender) [inline src/main/webapp/i18n/de/result.json:77]:
```suggestion
            "afterDueDateFeedbackHidden": "Auch nach der Abgabefrist werden Detailfeedbacks nicht zu allen Testfällen angezeigt. Dies ist etwa der Fall wenn für andere Studierende eine spätere Abgabefrist gesetzt ist (z.B. zum Ausgleich von Krankheitstagen). Alle Feedbacks werden automatisch am {{date}} sichtbar sein."
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java [hunk 2]
```diff
@@ -41,31 +41,47 @@ public class ProgrammingExerciseTestCaseResource {
...
+     * If the accessor is a student all invisible test cases are filtered and the hidden test cases
      *
-     * @param exerciseId of the the exercise.
+     * @param exerciseId of the exercise.
      * @return the found test cases or an empty list if no test cases were found.
      */
     @GetMapping(Endpoints.TEST_CASES)
-    @PreAuthorize("hasRole('TA')")
+    @PreAuthorize("hasRole('USER')")
     public ResponseEntity<Set<ProgrammingExerciseTestCase>> getTestCases(@PathVariable Long exerciseId) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java:68]: The test class for this resource did not change as far as I can see. Are there tests to make sure that this endpoint does not reveal tests unintentionally to students with the changed accessibility?

**Review by TheTrueVester — DISMISSED** [review]
Tested on ts1. Everything worked as expected.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java (hunk 2 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java]: This was a leftover because I tried to fix issue #4452 in this PR as well.
But I could not find an easy way without refactoring too much. So this will have to be postponed to a future PR.

The endpoint is now back to the version before this PR

**Review by TheTrueVester — DISMISSED** [review]
Tested on ts1. Only encountered one slight issue.
This is probably unrelated, but a score of 0% somehow filled up the score bar partially.
image

**Review by max-bergmann — APPROVED** [review]
Tested on ts2. Looks good

**Review by babyygemperor — APPROVED** [review]
Tested on ts2, works as described

**Review by rhubinger — DISMISSED** [review]
Tested on TS2. Everything works fine.

**Review by mallgaier — APPROVED** [review]
Tested on TS3, works as expected.
Code changes look good. 
Nice enhancement 👍

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 3]
```diff
@@ -215,6 +215,10 @@ public void setPositive(Boolean positive) {
         this.positive = positive;
     }
 
+    public void setPositiveViaCredits(Double credits) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:218]: Why do you have the credits as a parameter here? I think this should use the credits-attribute directly or update both attributes (credits and positive)

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 4]
```diff
@@ -323,4 +327,18 @@ public ResponseEntity<Boolean> toggleSecondCorrectionEnabled(@PathVariable Long
         return ResponseEntity.ok(exerciseRepository.toggleSecondCorrection(exercise));
     }
 
+    /**
+     * GET /exercises/{exerciseId}/latest-due-date
+     *
+     * @param exerciseId the exerciseId of the exercise to get the repos from
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:333]:
```suggestion
     * @param exerciseId the exerciseId of the exercise to get the last due date from
```

**Review by Strohgelaender — COMMENTED** [review]
I'll review the client code later.

File: src/main/webapp/app/exercises/shared/result/result-detail.component.html [hunk 5]
```diff
@@ -56,16 +58,26 @@ <h4 class="modal-title">Feedback</h4>
                 </div>
                 <div *ngFor="let feedback of filteredFeedbackList" [ngClass]="['feedback-item', 'alert', getClassNameForFeedbackItem(feedback)]" id="feedback-message">
                     <span>{{ feedback.category }}</span>
-                    <span id="feedback-points" class="feedback-points" *ngIf="feedback.credits"
-                        >{{ roundScoreSpecifiedByCourseSettings(feedback.credits, course) }}P<fa-icon
+                    <span id="feedback-points" class="feedback-points" *ngIf="feedback.credits">
+                        <ng-container *ngIf="feedback.actualCredits && feedback.actualCredits != feedback.credits">
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.html:62]:
```suggestion
                        <ng-container *ngIf="feedback.actualCredits && feedback.actualCredits !== feedback.credits">
```

File: src/main/webapp/app/exercises/shared/result/result-detail.component.html [hunk 6]
```diff
@@ -56,16 +58,26 @@ <h4 class="modal-title">Feedback</h4>
...
+                            {{ roundScoreSpecifiedByCourseSettings(feedback.actualCredits, course) }}P/
+                        </ng-container>
+                        {{ roundScoreSpecifiedByCourseSettings(feedback.credits, course) }}P
+                        <fa-icon
                             *ngIf="feedback.type === FeedbackItemType.Subsequent"
                             [icon]="faExclamationTriangle"
                             [ngbTooltip]="'artemisApp.assessment.subsequentFeedback' | artemisTranslate"
                         ></fa-icon>
+                        <fa-icon
+                            *ngIf="feedback.type === FeedbackItemType.Issue && feedback.actualCredits && feedback.actualCredits != feedback.credits"
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.html:72]: The second check (`feedback.actualCredits`) would evaluate to false for actualCredits == 0 (since 0 is falsy). I feel like this should be not the case here.
```suggestion
                            *ngIf="feedback.type === FeedbackItemType.Issue && feedback.actualCredits != null && feedback.actualCredits !== feedback.credits"
```

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 7]
```diff
@@ -341,7 +342,7 @@ export class ResultDetailComponent implements OnInit {
     private createProgrammingExerciseScaFeedbackItem(feedback: Feedback): FeedbackItem {
         const scaCategory = feedback.text!.substring(STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER.length);
         const scaIssue = StaticCodeAnalysisIssue.fromFeedback(feedback);
-        const text = this.showTestDetails ? `${scaIssue.rule}: ${scaIssue.message}` : scaIssue.message;
+        const text = `${scaIssue.rule}: ${scaIssue.message}`;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:345]: Was this change on purpose? If yes, why?

File: src/main/webapp/app/exercises/shared/result/result.component.ts [hunk 8]
```diff
@@ -518,4 +522,21 @@ export class ResultComponent implements OnInit, OnChanges {
...
+    private determineShowMissingAutomaticFeedbackInformation(componentInstance: ResultDetailComponent) {
+        if (!this.latestIndividualDueDate) {
+            this.exerciseService.getLatestDueDate(this.exercise!.id!).subscribe((latestIndividualDueDate?: dayjs.Dayjs) => {
+                this.latestIndividualDueDate = latestIndividualDueDate;
+                componentInstance.showMissingAutomaticFeedbackInformation = dayjs().isBefore(this.latestIndividualDueDate);
+                componentInstance.latestIndividualDueDate = this.latestIndividualDueDate;
+            });
+        } else {
+            componentInstance.showMissingAutomaticFeedbackInformation = dayjs().isBefore(this.latestIndividualDueDate);
+            componentInstance.latestIndividualDueDate = this.latestIndividualDueDate;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/result/result.component.ts:539]: Even if it's only two lines, we should move this into a method and not duplicate it.

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts (hunk 7 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/result/result-detail.component.ts:345]: This allows the students to actually see which SCA feedback belongs to which category. Since this PR now actually shows the amount of deducted points correctly if the student reaches the deduction limit, this information is necessary to see what other issues belong to the same category.
Beside that, this is no hidden information. Every student could deduct the SCA categories from the feedback if they want to.

Comment (Strohgelaender) [reply src/main/webapp/app/exercises/shared/result/result-detail.component.ts:345]: The category is not the rule here. You can see the difference (tutor view vs student view) below. 

grafik
grafik

I also think that some names could get confusing for the students: 
grafik

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 9]
```diff
@@ -215,6 +215,10 @@ public void setPositive(Boolean positive) {
         this.positive = positive;
     }
 
+    public void setPositiveViaCredits() {
+        this.positive = credits >= 0;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:219]:
```suggestion
        this.positive = credits != null && credits >= 0;
```
From the comment above the other change in this file, I assume that credits can actually be `null`.

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts (hunk 7 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/result/result-detail.component.ts:345]: You're right!

**Review by Strohgelaender — DISMISSED** [review]
code looks good

**Review by b-fein — APPROVED** [review]
Code looks good.

**Review by b-fein — CHANGES_REQUESTED** [review]
Revert approval: There are failing client tests on Bamboo that seem to be related to this PR (`ResultDetailComponent`).

**Comment by JohannesStoehr** [comment]
> Revert approval: There are failing client tests on Bamboo that seem to be related to this PR (`ResultDetailComponent`).

You were right. This was introduced by a change requested by Lucas. But the tests are also fixed (see last commit)

**Review by b-fein — APPROVED** [review]
Code looks good.

**Review by Dominik-Weinzierl — APPROVED** [review]
Code looks good to me :+1:

**Comment by JohannesStoehr** [comment]
> Tried to test it on TS2 with @yuling-sun . But we couldn't find how to create programming exercises with manual correction. How can you do so? Also, we aren't sure what is meant by submission policy. Do you mean for example lock repositories? Thanks ;)

@SimonPaul02 You can change the assessment in the timeline of the programming exercise after defining a deadline.
Lock Repository is one of the possible submission policies as indicated by the header "Submission Policy" right above this dropdown menu.

**Review by Dominik-Weinzierl — APPROVED** [review]
Tested in the testing session - works as expected :+1: Well done! Nice feature

**Review by ge65cer — APPROVED** [review]
Tested on TS1 in testing session, works alright.
_Please make sure the tests pass before merging (currently fail during dependency installation)._
